lib/ask/runtime/channel_broker_state.ex

```diff
@@ -80,201 +58,234 @@ defmodule Ask.Runtime.ChannelBrokerState do
+    )
   end
 
-  def get_channel_state(state, respondent_id) do
-    case Map.get(state.active_contacts, respondent_id) do
-      %{channel_state: channel_state} -> channel_state
-      # TODO: shall we raise instead?
-      _ -> %{}
-    end
+  # Returns true if a respondent is currently in queue (active or not).
+  def queued_or_active?(state, respondent_id) do
+    Repo.exists?(
+      from q in Queue,
+        where: q.channel_id == ^state.channel_id and q.respondent_id == ^respondent_id
+    )
   end
 
   # Adds a contact to the queue. The priority is set from the respondent's
   # disposition.
   def queue_contact(state, contact, size) do
     respondent = elem(contact, 0)
-    priority = if respondent.disposition == :queued, do: :normal, else: :high
-    queue_contact(state, contact, size, priority)
+
+    if respondent.disposition == :queued do
+      queue_contact(state, contact, size, :normal)
+    else
+      queue_contact(state, contact, size, :high)
+    end
   end
 
-  # Adds a contact to the queue with given priority (`:high`, `:normal`, `:low`).
-  def queue_contact(state, contact, size, priority) do
-    new_contacts_queue = PQueue.push(state.contacts_queue, [size, contact], priority)
-    Map.put(state, :contacts_queue, new_contacts_queue)
+  # Adds an IVR contact to the queue with given priority (`:high`, `:normal`, `:low`).
+  def queue_contact(state, {respondent, token, not_before, not_after}, size, priority) do
+    Queue.upsert!(%{
+      channel_id: state.channel_id,
+      respondent_id: respondent.id,
+      queued_at: SystemTime.time().now,
+      priority: priority,
+      size: size,
+      token: token,
+      not_before: not_before,
+      not_after: not_after,
+      reply: nil
+    })
+
+    state
   end
 
-  # Updates the active contact for the respondent. Does nothing if there are
-  # no active contact for this respondent.
-  defp update_active_contact(%{active_contacts: active_contacts} = state, respondent_id, cb) do
-    if active_contact = Map.get(active_contacts, respondent_id) do
-      new_active_contacts = Map.put(active_contacts, respondent_id, cb.(active_contact))
-      Map.put(state, :active_contacts, new_active_contacts)
-    else
-      state
-    end
+  # Adds an SMS contact to the queue with given priority (`:high`, `:normal`, `:low`).
+  def queue_contact(state, {respondent, token, reply}, size, priority) do
+    Queue.upsert!(%{
+      channel_id: state.channel_id,
+      respondent_id: respondent.id,
+      queued_at: SystemTime.time().now,
+      priority: priority,
+      size: size,
+      token: token,
+      not_before: nil,
+      not_after: nil,
+      reply: reply
+    })
+
+    state
+  end
+
+  def put_channel_state(state, respondent_id, channel_state) do
+    query =
+      from q in Queue,
+        where: q.channel_id == ^state.channel_id and q.respondent_id == ^respondent_id
+
+    Repo.update_all(query, set: [channel_state: channel_state])
```

Comment:
Hum? This is already a single UPDATE. `query` is just the pending SQL statement.

Or do you mean the overall logic:

1. queue (upsert)
2. activate (update × 1) -> before request to verboice/nuntium
3. put channel state (update × 2) -> after request to verboice/nuntium
4. each callback (update × n)
5. final callback (delete)

Reply:
I was refering to the `Repo.update_all` instead of doing a `Repo.update` or adding a `LIMIT 1`.

Reply:
Oh! I use `Repo.update_all()` for a direct SQL request, because `Repo.all` requires a changeset, and thus require to load -> changeset -> update. I could had `LIMIT 1` but we're already updating by the primary key (composite channel_id, respondent_id), so there will only be one row at most.